api-examples/transformer_utils.py

```diff
@@ -0,0 +1,422 @@
+                    if response:
+                        yield response
+
+    def process_line(self, line):
+        """Read in a line and turn it into an entry
+
+        The entries will get collated by the data loader
+
+        :param line:
+        :return:
+        """
+
+
+class NextTurnPredictionFileLoader(MultiFileLoader):
+
+    def process_line(self, line):
+        pair = line.strip().split('\t')
+        # Unfortunately, this occassionally happens, a bunch of blank turns etc.
+        if len(pair) != 2:
+            return None
+        q, r = pair
+        if q == '' or r == '':
+            return None
+        q_vec, q_valid_lengths = self.vectorizer.run(reversed(q.split()), self.vocab)
+        q_vec = np.roll(q_vec[::-1], -(self.vectorizer.mxlen - q_valid_lengths))
+        r_vec, r_valid_lengths = self.vectorizer.run(r.split(), self.vocab)
+        return q_vec, r_vec
+
+
+class SequencePredictionFileLoader(MultiFileLoader):
+
+    def process_line(self, line):
+        line = line.strip()
+        if not line:
+            return None
+
+        vec, valid_lengths = self.vectorizer.run(reversed(line.split()), self.vocab)
+        if valid_lengths < 2:
+            return None
+        vec = np.roll(vec[::-1], -(self.vectorizer.mxlen - valid_lengths))
+        return vec, vec
+
+
+class NextSequencePredictionFileLoader(MultiFileLoader):
+
+    def process_line(self, line):
+        line = line.strip()
+        if not line:
+            return None
+        vec, valid_lengths = self.vectorizer.run(reversed(line.split()), self.vocab)
+        vec = np.roll(vec[::-1], -(self.vectorizer.mxlen - valid_lengths))
+        if valid_lengths < 2:
+            return None
+        half_lengths = self.vectorizer.mxlen//2
+        context = vec[:half_lengths]
+        response = vec[half_lengths:]
+        return context, response
+
+
+class MultiFileDatasetReader:
+    """Provide a base-class to do operations that are independent of token representation
+    """
+
+    def __init__(self, nctx=64, model_file=None, vocab_file=None, pattern='*.txt', reader_type="ntp"):
+        self.nctx = nctx
+        self.pattern = pattern
+        self.reader_type = reader_type
+        self.vectorizer = BPEVectorizer1D(model_file=model_file, vocab_file=vocab_file, mxlen=nctx)
+
+    def build_vocab(self, _=None):
+        return {'x': self.vectorizer.vocab}
+
+    def load(self, directory, vocabs):
+        reader_type = self.reader_type.lower()
+        if reader_type == "ntp":
+            return NextTurnPredictionFileLoader(directory, self.pattern, vocabs, self.vectorizer, self.nctx)
+        elif reader_type == "nsp":
+            return NextSequencePredictionFileLoader(directory, self.pattern, vocabs, self.vectorizer, 2*self.nctx)
+        else:
+            print("Using files as an LM")
+            return SequencePredictionFileLoader(directory, self.pattern, vocabs, self.vectorizer, self.nctx)
+
+
+class TiedSeq2SeqModel(Seq2SeqModel):
```

Comment:
What is tied here? just the embeddings? The name makes me think it is doing weight sharing between the Encoder and Decoder but I don't see it forcing that unless I am missing something

Reply:
yes, this could have been clearer, it differs from the base class in that it ties the embeddings